rfd/0180-sso-as-mfa-method.md

```diff
@@ -0,0 +1,767 @@
+re-authenticate through their SSO provider for a more seamless experience.
+
+```console
+> tsh ssh server01
+MFA is required to access Node "server01"
+Complete an auth check in your local web browser:
+If browser window does not open automatically, open it by clicking on the link:
+ http://127.0.0.1:60433/f5858c78-75e1-4f3f-b2c5-69e8e76c0ff9
+```
+
+##### Adding first MFA Device
+
+Usually, a new SSO user without a registered MFA device can add their first
+device without an additional authentication check. With SSO MFA, we can include
+an MFA check even for the first device.
+
+TODO:: show tsh add and delete
+
+#### Existing SSO user with registered MFA method
+
+> I have logged into this cluster before with SSO
+> I have registered one or more MFA devices
+> I want to connect to a resource protected by per-session MFA
+
+##### SSO MFA prompt
+
+The user will be given the option to pass MFA checks with a registered device
+or with SSO, depending on the MFA mode of the user's auth connector.
+
+```console
+> tsh ssh server01
+MFA is required to access Node "server01"
+Complete an auth check in your local web browser. Open it by clicking on the link:
+ http://127.0.0.1:60433/f5858c78-75e1-4f3f-b2c5-69e8e76c0ff9
+ 
+Or tap any security key
+```
+
+Note that the exact prompt may differ depending on which [config scenario](#tsh)
+we are in.
+
+##### Deleting last MFA device
+
+Similar to how we can add an SSO MFA check for a [user adding their first device](#adding-first-mfa-device),
+we can use an SSO MFA check for users removing their last device, which is
+usually forbidden. This may be useful if the user wants to makes SSO their
+default MFA method for UX purposes.
+
+### Configuration
+
+#### Enable SSO connector as MFA method.
+
+Auth connectors will have a new `mfa` settings field that can be set to allow
+an auth connector to handle MFA checks in addition to login requests. These
+MFA settings will apply to all users registered through the Auth connector.
+
+```yaml
+kind: saml
+version: v2
+metadata:
+  name: okta
+spec:
+  display: Okta
+  acs: https://root.example.com:3080/v1/webapi/saml/acs/okta
+  entity_descriptor_url: AAA
+  entity_descriptor: BBB
+  attributes_to_roles:
+    - {name: groups, value: okta-dev, roles: [dev]}
+  # new mfa settings field. Defaults to null.
+  mfa:
+    # enabled specifies whether this auth connector supports MFA checks.
+    enabled: yes
+    # mode can be set to determine how Teleport decides to prioritize SSO MFA
+    # over other MFA options for users tied to this auth connector. Supported
+    # values are optional (default), preferred, and required.
+    mode: optional
+    # set entity_descriptor_url or entity_descriptor to use a different IdP configured
+    # app to handle MFA checks. This is useful when configuring a separate MFA flow
+    # tied to a separate app. Defaults to the entity_descriptor_url and entity_descriptor above.
+    entity_descriptor_url: XXX
```

Comment:
> Since this is the same IdP, I am not seeing additional benefit of supporting MFA specific IdP configuration except for new mfa: {enabled, mode, force_auth} knobs.

In the intended use case, an IdP admin would create a second service application with a different login flow: e.g. take the user's existing IdP login session and prompt them for MFA.

There are other cases where an admin might want to use the same IdP service application and trigger a different MFA flow with a different `prompt` or `acr_values`, etc. 

And the final use case we support is to use the same IdP service application for both login and MFA. Users would be going through the same exact login flow instead of a specialized MFA flow. This is the least optimal use case, so I think supporting it by default by defaulting the client/entity_descriptor fields to the parent connector was actually a mistake, we should just require they are set even if duplicating the fields is a bit "ugly".

> It probably wont be a situation where admin less securely configures the auth connector and more securely configures the MFA connector or vice versa?

Actually no, this would be expected.

As an example, a user's IdP login session could last 8 hours. Meaning if they re-login through teleport within 8 hours, they can refresh their Teleport login session without resupplying their login credentials. However, if the user requests an MFA protected action through Teleport, we always want to check their MFA credentials. We don't necessarily need them to provide their username and password since they have an active seession, but we do need them to re-supply MFA.

It's possible to configure this flow through a single app but whether that's easy or difficult would depend on the IdP in question. In most cases it's probably easier to configure a second app.